src/client.js

```diff
@@ -217,72 +227,60 @@ assign(Client.prototype, {
       helpers.warn('The pool has already been initialized')
       return
     }
-    this.pool = new Pool(assign(this.poolDefaults(config.pool || {}), config.pool))
+
+    const poolSettings = this.getPoolSettings(config.pool);
+
+    this.pool = genericPool.createPool(poolSettings.factory, poolSettings.config)
   },
 
   validateConnection(connection) {
-    return true
+    return Promise.resolve(true);
   },
 
   // Acquire a connection from the pool.
   acquireConnection() {
-    return new Promise((resolver, rejecter) => {
-      if (!this.pool) {
-        return rejecter(new Error('Unable to acquire a connection'))
-      }
-      let wasRejected = false
-      const t = setTimeout(() => {
-        wasRejected = true
-        rejecter(new Promise.TimeoutError(
+    if (!this.pool) {
+      return Promise.reject(new Error('Unable to acquire a connection'))
+    }
+    return this.pool.acquire()
+      .tap(connection => {
+        debug('acquired connection from pool: %s', connection.__knexUid)
+      })
+      .catch(genericPoolErrors.TimeoutError, () => {
+        throw new Promise.TimeoutError(
           'Knex: Timeout acquiring a connection. The pool is probably full. ' +
           'Are you missing a .transacting(trx) call?'
-        ))
-      }, this.config.acquireConnectionTimeout || 60000)
-      this.pool.acquire((err, connection) => {
-        clearTimeout(t)
-        if (err) {
-          return rejecter(err)
-        }
-        if (wasRejected) {
-          this.pool.release(connection)
-        } else {
-          debug('acquired connection from pool: %s', connection.__knexUid)
-          resolver(connection)
-        }
-      })
-    })
+        )
+      });
   },
 
   // Releases a connection back to the connection pool,
   // returning a promise resolved when the connection is released.
   releaseConnection(connection) {
-    return new Promise((resolver) => {
-      debug('releasing connection to pool: %s', connection.__knexUid)
-      this.pool.release(connection)
-      resolver()
+    debug('releasing connection to pool: %s', connection.__knexUid)
+    return this.pool.release(connection).catch(() => {
+      debug('pool refused connection: %s', connection.__knexUid)
     })
   },
 
   // Destroy the current connection pool for the client.
   destroy(callback) {
-    const promise = new Promise((resolver) => {
+    return new Promise((resolver) => {
```

Comment:
could this be written:

```
return Promise.resolve(
  this.pool && 
  this.pool.drain()
    .then(() => this.pool.clear())
    .then(() => {
      this.pool = void 0; 
      if(typeof callback === 'function') {
        callback();
      }
    })
);
```

to prevent need for new Promise()  + remembering to call resolve function.